src/Monad/Collection.php

```diff
@@ -45,20 +45,20 @@ public function map(callable $transformation)
      */
     public function ap(FantasyLand\Apply $applicative)
     {
-        // Sine in php List comprehension is not available, then I doing it like this
+        // Since we don't have List comprehension in PHP, use a foreach
         $result = [];
         $isCollection = $applicative instanceof Collection;
 
         foreach ($this->extract() as $value) {
-            $partial = $applicative->map($value)->extract();
+            $partial = f\valueOf($applicative->map($value));
             if ($isCollection) {
                 $result = f\push($result, $partial);
             } else {
                 $result[] = $partial;
             }
         }
 
-        return $applicative::of($result);
+        return $applicative instanceof FantasyLand\Pointed ? $applicative::of($result) : $result;
```

Comment:
This is not needed -`Applicative` inherits `Pointed`.
https://github.com/widmogrod/php-functional/blob/master/src/FantasyLand/Applicative.php